feat(distribution): add fee-distribution and avs-reward distribution

mikebraver opened the pull request:
## Description
Add the distribution module logic and minting reward related.
The distribution mainly contains two parts:

1. The fee distribution to exocore validator and corresponding stakers, which come from the minting rewards.
2.  The reward allocation/distribution to avs, which is based on the commission of avs params as well as reward calculation for avs.
And generally, these will be allocated on epoch-basis, say, one day. And upon the epoch end, all these two hooks will be triggered to distribute the reward.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

----


<!--
< < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

PR review checkboxes:

I have...

- [ ] added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] included the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] targeted the correct branch
      (see [PR Targeting](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link in the PR description to the relevant issue or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all required CI checks have passed

Code maintenance:

I have...

- [ ] written unit and integration [tests](https://github.com/ExocoreNetwork/exocore/blob/main/CONTRIBUTING.md#testing)
- [ ] added relevant [`godoc`](https://go.dev/blog/godoc) and [code comments](https://blog.jbowen.dev/2019/09/the-magic-of-go-comments/).
- [ ] updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)

______

### Reviewers Checklist

**All** items are required.
Please add a note if the item is not applicable
and please add your handle next to the items reviewed
if you only reviewed selected items.

I have...

- [ ] confirmed the correct
      [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
      in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] confirmed that this PR does not change production code

-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive token allocation mechanism for fair reward distribution among validators and stakers in the blockchain ecosystem.
  - Enhanced commission structure for operators, increasing commission rates for improved financial dynamics.
  - Added new methods for improved interaction with operators, stakers, and AVS assets.

- **Bug Fixes**
  - Improved reward distribution accuracy by refining allocation logic and adding necessary validation checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

mikebraver committed:
feat: initialize distribution

mikebraver committed:
update

mikebraver committed:
feat: add distribution logic

update

mikebraver committed:
feat: initialize fee distribution framework

mikebraver committed:
feat: add the avs reward

mikebraver committed:
feat: add the fee distribution allocation

mikebraver committed:
fix: some fixes

leonz789 commented on proto/exocore/avs/tx.proto:
```diff
@@ -28,14 +28,16 @@ message AVSInfo {
   // and current_epoch is the integer identifier of the epoch module
   int64 starting_epoch = 10;
+  // asset_reward_commission_epoch_basis is the avs reward distribution based on asset per eopch end.
+  map<string, int64> asset_reward_amount_epoch_basis = 11;
```
map in message might cause error when parse within cli, better to change to list(repeated).

leonz789 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,127 @@
+		if err := k.FeePool.Set(ctx, types.FeePool{CommunityPool: feePool.CommunityPool.Add(feesCollected...)}); err != nil {
+			return err
+		}
+	}
```
If all feesCollected are added to the communityPool, should the distribution continue or just return ? (set the remaining logic as a `else` case, or just return when `totalPreviousPower==0`)

mikebraver replied:
For the first time, which is the case that the totalPreviousPower is 0, we can skip the distribution and let it happen next time the power occur?

leonz789 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,127 @@
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCommission,
+		sdk.NewAttribute(sdk.AttributeKeyAmount, commission.String()),
+		sdk.NewAttribute(types.EventTypeCommission, val.GetOperator().String()),
```
keyvalidator ?

leonz789 commented on x/feedistribution/keeper/keeper.go:
```diff
@@ -0,0 +1,74 @@
+	}
+)
+
+func NewKeeper(
```
collections need to be initialized with module's store

mikebraver replied:
updated to avoid collections for compatible

leonz789 commented on x/feedistribution/keeper/hooks.go:
```diff
@@ -0,0 +1,17 @@
+package keeper
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
```
need hooks to track delegation/undelegation for f1 period update

leonz789 commented on x/feedistribution/client/cli/tx.go:
```diff
@@ -0,0 +1,20 @@
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/spf13/cobra"
+)
+
```
cli for delegators to withdraw rewards

mikebraver committed:
fix: fix the dependency and test errors

mikebraver committed:
resolve: fix merge conflicts

resolve: fix merge conflicts

up

mikebraver committed:
chore: some refactors

mikebraver committed:
fix: fix all test errors

mikebraver committed:
chore: refactor the protobuf directory path

MaxMustermann2 commented on proto/exocore/feedistribute/v1/distribution.proto:
```diff
@@ -0,0 +1,80 @@
+    (gogoproto.nullable) = false,
+    (amino.dont_omitempty) = true,
+    (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.DecCoins",
+    deprecated = true
```
Why add a deprecated field?

MaxMustermann2 commented on proto/exocore/feedistribute/v1/params.proto:
```diff
@@ -0,0 +1,22 @@
+message Params {
+  option (amino.name) = "exocore/x/feedistribution/Params";
+  // epoch_identifier for fee distribution
+  string epoch_identifier = 2;
```
Start from 1?

mikebraver replied:
As long as the number is in order, it's fine, right?

MaxMustermann2 replied:
It is unconventional but functionally it is the same.

MaxMustermann2 commented on proto/exocore/feedistribute/v1/tx.proto:
```diff
@@ -0,0 +1,40 @@
+
+  // UpdateParams defines a (governance) operation for updating the module
+  // parameters. The authority defaults to the x/gov module account.
+  rpc UpdateParams(MsgUpdateParams) returns (MsgUpdateParamsResponse);
```
Add `option (google.api.http).post` ?

MaxMustermann2 commented on x/feedistribution/client/cli/query.go:
```diff
@@ -0,0 +1,48 @@
+	return cmd
+}
+
+func QueryDistritbutionInfo() *cobra.Command {
```
Typo: `QueryDistributionInfo`

MaxMustermann2 commented on x/feedistribution/client/cli/tx.go:
```diff
@@ -0,0 +1,72 @@
+			if err != nil {
+				return err
+			}
+			sender := cliCtx.GetFromAddress()
```
Consider calling `epochstypes.ValidateEpochIdentifierString` on `args[0]`

trestinlsd commented on x/feedistribution/client/cli/tx.go:
```diff
@@ -0,0 +1,72 @@
+// CmdUpdateParams is to update Params for distribution module
+func CmdUpdateParams() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "update-params",
```
Explain the use of parameters：Use: update-params <EpochIdentifier>  <communityInteger> <communityPrecise>

trestinlsd replied:
Use:   "TaskInfo <EpochIdentifier> <communityInteger> <communityPrecise>

mikebraver committed:
some fixes

trestinlsd commented on x/feedistribution/keeper/epoch_info.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+// SetEpochsPerPeriod stores the epoch identifier
+func (k Keeper) SetEpochIdentifier(ctx sdk.Context, epochIdentifier string) {
```
This method was not used

mikebraver committed:
fix: fix rebase conflict

MaxMustermann2 commented on x/delegation/keeper/delegation_op_test.go:
```diff
@@ -145,13 +147,15 @@ func (suite *DelegationTestSuite) TestUndelegateFrom() {
 	}, *operatorState)
-
+	if !ok {
+		log.Error("not ok")
```
Instead of `Equal`, maybe `Assert` ?

TimmyExogenous replied:
Is this modification being made because the test fails without any related code changes? If so, I can record this issue and look into the cause when I have time later.

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,92 @@
+	// allocate tokens proportionally to voting power of different validators
+	//
+	// TODO: Consider parallelizing later
+	validatorUpdates := k.StakingKeeper.GetValidatorUpdates(ctx)
```
Isn't this the `new` voting power? And you're using it to allocate past rewards?

MaxMustermann2 commented on x/feedistribution/types/genesis.go:
```diff
@@ -0,0 +1,24 @@
+)
+
+// DefaultIndex is the default global index
+const DefaultIndex uint64 = 1
```
unused.

MaxMustermann2 commented on x/feedistribution/types/params.go:
```diff
@@ -0,0 +1,41 @@
+	}
+}
+
+// Validate validates the set of params
+func (p Params) Validate() error {
+	return nil
+}
```
Check if commission is negative?

MaxMustermann2 replied:
And also just call the `epochstypes.ValidatorEpochIdentifierString` function

MaxMustermann2 replied:
Sorry my bad it is called `ValidateEpochIdentifierString`

MaxMustermann2 commented on x/reward/keeper/hooks.go:
```diff
@@ -0,0 +1,93 @@
+				}
+				if wrapper.keeper.assetsKeeper.IsOperatorAssetExist(ctx, opAddr, asset) {
+					coin := sdk.Coin{
+						Denom:  assetInfo.AssetBasicInfo.Symbol,
```
If the AVS is based on `ETH`, this will mint them (fake) `ETH` on Cosmos. Is that intentional (for some testing) ?

mikebraver replied:
In such case, as avs would second-custom their wanted assets, and if the avs developer defined `ETH` as reward, then we record this reward recording

MaxMustermann2 commented on x/reward/keeper/keeper.go:
```diff
@@ -5,33 +5,81 @@ import (
+	if !store.Has(poolKey.Bytes()) {
+		return newRewardRecord(ctx, k, k.banker, k.distributor, types.NewPool(name))
+	}
+	return newRewardRecord(ctx, k, k.banker, k.distributor, pool)
```
`pool` is nil here?

MaxMustermann2 commented on x/reward/keeper/reward_record.go:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	for i, reward := range p.Rewards {
+		if reward.EarningsAddr == earningAddress {
```
Is there a better way than looping over possibly multiple operators?

trestinlsd committed:
fix:fix consensus address parse error

trestinlsd committed:
Add test dependencies and initialize genesis

trestinlsd committed:
fix:ix distribution keep startup sequence

trestinlsd committed:
fix:fix stored fee pool should not have been nil

mikebraver committed:
update

mikebraver committed:
feat: add reward allocation to stakers

mikebraver committed:
fix: some trivial fix

mikebraver committed:
test: add more uts

MaxMustermann2 commented on x/dogfood/keeper/keeper.go:
```diff
@@ -114,3 +111,24 @@ func (k Keeper) mustValidateFields() {
 }
+
+// Add the function to get detail information through the operatorKeeper within the dogfood
+func (k Keeper) ValidatorByConsAddrForChainID(ctx sdk.Context, consAddr sdk.ConsAddress, chainID string) (stakingtypes.Validator, bool) {
```
Why put this function here? From the context of `x/dogfood` there is only one `chainID`: the `ctx.ChainID()`.

I know we discussed this approach, but why can't these methods be called on the operator keeper instead? The distribution keeper's constructor can be initialized with it.

mikebraver replied:
As the distribution keeper needs the dogfood staking keeper, and operator keeper is already inside the staking keeper, so it seemed redundant to add operator keeper in the distribution constructor

MaxMustermann2 replied:
I don't like this approach. Cleaner to add operator keeper in the distribution constructor. But not a blocker to merge.

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,148 @@
+	logger.Info("Allocate tokens to validator successfully", "allocated amount is", outstanding.Rewards.String())
+}
+
+func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.ValAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
```
This function is too heavy. But I don't have any ideas to improve it either.

mikebraver replied:
For our current approach, it's acceptable to realize the feature, and we'll optimize it afterwards

MaxMustermann2 commented on x/reward/keeper/hooks.go:
```diff
@@ -0,0 +1,93 @@
+) {
+	expEpochIdentifier := wrapper.keeper.GetEpochIdentifier(ctx)
+	if epochIdentifier != expEpochIdentifier {
+		wrapper.keeper.Logger(ctx).Error(
```
This log is not necessary. Maybe reduce the level to Debug.

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,148 @@
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotoalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
```
typo: `curTotalStakersPowers`

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,148 @@
+		powerFraction := stakerPower.QuoTruncate(curTotoalStakersPowers)
+		rewardToSingleStaker := rewardToAllStakers.MulDecTruncate(powerFraction)
+		k.AllocateTokensToSingleStaker(ctx, staker, rewardToSingleStaker)
+		rewardToAllStakers = rewardToAllStakers.Sub(rewardToSingleStaker)
```
Safety check that `rewardToAllStakers` isn't negative somehow? Or it is not needed?

mikebraver replied:
It'll not be negative. As the deduction is based on the weight ratio of different stakers

bwhour commented on x/dogfood/keeper/keeper.go:
```diff
@@ -114,3 +111,24 @@ func (k Keeper) mustValidateFields() {
+	return k.operatorKeeper.ValidatorByConsAddrForChainID(ctx, consAddr, chainID)
+}
+
+func (k *Keeper) GetStakersByOperator(ctx sdk.Context, operator, assetID string) (delegationtype.StakerList, error) {
```
only this function is  pointer receiver, is it specail ?

mikebraver replied:
For compatible with other modules

TimmyExogenous commented on x/reward/keeper/hooks.go:
```diff
@@ -0,0 +1,93 @@
+			)
+			return
+		}
+		assetID := avsInfo.Info.AssetIDs
```
The `AssetIDs` seem to be the list of assets supported by the AVS. Could you please explain why the reward token depends on this assets list?

mikebraver replied:
The reward will not be much enabled for now as discussed. Can pay more attention to allocation in this pr

TimmyExogenous commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,148 @@
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotoalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
+	for _, avsAddress := range avsList {
```
Could we pass the AVS as a parameter to handle this?

mikebraver replied:
Seemed no difference

cloud8little commented:
Test failed, running local_node.sh to start a single validator network, check distribution logs, there is no logs for single staker, it is due to the operatorAddress used is not Bench32 address. The code use `operatorAddress.String()` , use log to print it out, the result is `exovaloper18cggcpvwspnd5c6ny8wrqxpffj5zmhkl0le8yd`, it should be `exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph`

```
10:06AM INF Allocate tokens to all validators feesCollected amount is =20.000000000000000000aexo module=x/distribution
10:06AM INF avsList is=[] module=x/distribution
10:06AM INF operatorAddress is=exovaloper18cggcpvwspnd5c6ny8wrqxpffj5zmhkl0le8yd module=x/distribution
10:06AM INF staker map is={} module=x/distribution
10:06AM INF allocate tokens to stakers successfully allocated amount is=19.400000000000000000aexo module=x/distribution
10:06AM INF Allocate tokens to validator successfully allocated amount is=19.400000000000000000aexo module=x/distribution
```

```go
func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.ValAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
	logger := k.Logger()
	avsList, err := k.StakingKeeper.GetOptedInAVSForOperator(ctx, operatorAddress.String())
	if err != nil {
		ctx.Logger().Error("avs address lists not found; skipping")
	}
	stakersPowerMap, curTotoalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
	for _, avsAddress := range avsList {
		avsAssets, err := k.StakingKeeper.GetAVSSupportedAssets(ctx, avsAddress)
		if err != nil {
			ctx.Logger().Error("avs address lists not found; skipping")
		}
		for assetID := range avsAssets {
			stakerList, err := k.StakingKeeper.GetStakersByOperator(ctx, operatorAddress.String(), assetID)
			if err != nil {
				ctx.Logger().Error("staker lists not found; skipping")
			}
			for _, staker := range stakerList.Stakers {
				if curStakerPower, err := k.StakingKeeper.CalculateUSDValueForStaker(ctx, staker, avsAddress, operatorAddress.Bytes()); err != nil {
					ctx.Logger().Error("curStakerPower error", err)
				} else {
					stakersPowerMap[staker] = curStakerPower
					curTotoalStakersPowers.Add(curStakerPower)
				}
			}
		}
	}

	for staker, stakerPower := range stakersPowerMap {
		powerFraction := stakerPower.QuoTruncate(curTotoalStakersPowers)
		rewardToSingleStaker := rewardToAllStakers.MulDecTruncate(powerFraction)
		k.AllocateTokensToSingleStaker(ctx, staker, rewardToSingleStaker)
		rewardToAllStakers = rewardToAllStakers.Sub(rewardToSingleStaker)
	}
	feePool.CommunityPool = feePool.CommunityPool.Add(rewardToAllStakers...)
	logger.Info("allocate tokens to stakers successfully", "allocated amount is", rewardToAllStakers.String())
}
```

mikebraver committed:
fix: fix the operator address format

MaxMustermann2 commented on x/feedistribution/keeper/msg_update_params.go:
```diff
@@ -0,0 +1,21 @@
+	// }
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	k.SetParams(ctx, req.Params)
```
Check that the `EpochIdentifier` is registered in the `x/epochs` module

mikebraver replied:
Distribution module already registered in epoch within app.go, and as well as `epochIdentifier`

MaxMustermann2 replied:
Not sure I understand your response. The `MsgUpdateParams` could contain an invalid epoch not registered within the `x/epochs` module. For example, change of the epoch identifier to `max` should be deemed invalid.

mikebraver committed:
resolve: resolve comments

MaxMustermann2 commented on app/app.go:
```diff
@@ -688,6 +697,19 @@ func NewExocoreApp(
+	// and it'll interact with other modules, like delegation for voting power, mint and inflation and etc.
+	// this keeper is initialized after the StakingKeeper  because it depends on the StakingKeeper
+	app.DistrKeeper = distrkeeper.NewKeeper(
+		appCodec, logger,
```
Why pass the logger here? Instead use `ctx.Logger()` where required

MaxMustermann2 commented on app/app.go:
```diff
@@ -966,6 +991,7 @@ func NewExocoreApp(
 		avsManagerTypes.ModuleName,
 		// op module
 		feemarkettypes.ModuleName, // last in order to retrieve the block gas used
+		distrtypes.ModuleName,
```
Please move this above the set of `no-op` modules and add comments about the order. Same for `InitGenesis` and `BeginBlockers`

MaxMustermann2 commented on proto/exocore/feedistribute/v1/distribution.proto:
```diff
@@ -0,0 +1,90 @@
+syntax = "proto3";
```
These files are in subfolder `feedistribute` but the generated `go` files are in `feedistribution` (same as the module). Please fix for consistency.

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+	for i, val := range allValidators {
+		pk, err := val.ConsPubKey()
+		if err != nil {
+			ctx.Logger().Error("Failed to deserialize public key; skipping", "error", err, "i", i)
```
```suggestion
			logger.Error("Failed to deserialize public key; skipping", "error", err, "i", i)
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+			ctx, sdk.GetConsAddress(pk), avstypes.ChainIDWithoutRevision(ctx.ChainID()),
+		)
+		if !found {
+			ctx.Logger().Error("Operator address not found; skipping", "consAddress", sdk.GetConsAddress(pk), "i", i)
```
```suggestion
			logger.Error("Operator address not found; skipping", "consAddress", sdk.GetConsAddress(pk), "i", i)
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+	logger.Info("Allocate to stakers of operatorAddress:", operatorAddress.String())
+	avsList, err := k.StakingKeeper.GetOptedInAVSForOperator(ctx, operatorAddress.String())
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
```
```suggestion
		logger.Error("avs address lists not found; skipping")
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+	for _, avsAddress := range avsList {
+		avsAssets, err := k.StakingKeeper.GetAVSSupportedAssets(ctx, avsAddress)
+		if err != nil {
+			ctx.Logger().Error("avs address lists not found; skipping")
```
```suggestion
			logger.Error("avs address lists not found; skipping")
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+		for assetID := range avsAssets {
+			stakerList, err := k.StakingKeeper.GetStakersByOperator(ctx, operatorAddress.String(), assetID)
+			if err != nil {
+				ctx.Logger().Error("staker lists not found; skipping")
```
```suggestion
				logger.Error("staker lists not found; skipping")
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+			}
+			for _, staker := range stakerList.Stakers {
+				if curStakerPower, err := k.StakingKeeper.CalculateUSDValueForStaker(ctx, staker, avsAddress, operatorAddress.Bytes()); err != nil {
+					ctx.Logger().Error("curStakerPower error", err)
```
```suggestion
					logger.Error("Could not get staker power", "error", err)
```

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+
+func (k Keeper) AllocateTokensToStakers(ctx sdk.Context, operatorAddress sdk.AccAddress, rewardToAllStakers sdk.DecCoins, feePool *types.FeePool) {
+	logger := k.Logger()
+	logger.Info("Allocate to stakers of operatorAddress:", operatorAddress.String())
```
Follow interface `(msg string, keyvals ...interface{})`
```suggestion
	logger.Info("AllocateTokensToStakers", "operatorAddress", operatorAddress.String())
```

MaxMustermann2 commented on x/feedistribution/types/keys.go:
```diff
@@ -0,0 +1,84 @@
+
+const (
+	// ModuleName defines the module name
+	ModuleName = "distribution"
```
```suggestion
	ModuleName = "feedistribution"
```

MaxMustermann2 commented on x/feedistribution/types/keys.go:
```diff
@@ -0,0 +1,84 @@
+	RouterKey = ModuleName
+
+	// MemStoreKey defines the in-memory store key
+	MemStoreKey            = "mem_feedistribute"
```
```suggestion
	MemStoreKey            = "mem_feedistribution"
```

MaxMustermann2 commented on x/feedistribution/types/keys.go:
```diff
@@ -0,0 +1,84 @@
+
+	// MemStoreKey defines the in-memory store key
+	MemStoreKey            = "mem_feedistribute"
+	ProtocolPoolModuleName = "protocolpool"
```
This looks to be unused

MaxMustermann2 commented on x/feedistribution/types/params.go:
```diff
@@ -0,0 +1,44 @@
+	return NewParams()
+}
+
+// ParamSetPairs get the params.ParamSet
+func (p *Params) ParamSetPairs() paramtypes.ParamSetPairs {
+	return paramtypes.ParamSetPairs{}
+}
```
This feature is not required since the `x/params` module is deprecated. We should remove it.

cloud8little commented:
Test failed on https://github.com/ExocoreNetwork/exocore/pull/117/commits/26e1833f1869c3056ce014295db748dd723e22e7: 
consensus failure after price is feeding. notice when total exomint for each epoch is only 20 aexo, but allocate tokens to all validators is `18079023200020.000000000000000000aexo`. 

![1724639450229](https://github.com/user-attachments/assets/c90b7838-d776-4a25-a84d-9b8133580cfc)

```
2:14AM INF Allocate tokens to all validators feesCollected amount is =18079023200020.000000000000000000aexo module=x/distribution
2:14AM INF Allocate to stakers of operatorAddress: module=x/distribution
2:14AM INF allocate tokens to single staker successfully allocated amount is=240591473531018555390.000000000000000000aexo module=x/distribution
2:14AM ERR CONSENSUS FAILURE!!! err="negative coin amount" module=consensus server=node stack="goroutine 169 [running]:\nruntime/debug.Stack()\n\truntime/debug/stack.go:24 +0x5e\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine.func2()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:736 +0x46\npanic({0x2cad020?, 0x41a9420?})\n\truntime/panic.go:914 +0x21f\ngithub.com/cosmos/cosmos-sdk/types.DecCoins.Sub(...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.8/types/dec_coin.go:310\ngithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper.Keeper.AllocateTokensToStakers({{0x42050a0, 0xc000f30da0}, {0x41c2830, 0xc0014bf3f0}, {0x41ea158, 0xc001472b00}, {0xc001b5bb00, 0x2a}, {0x41ea318, 0xc0012d4730}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper/allocation.go:139 +0x73e\ngithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper.Keeper.AllocateTokensToValidator({{0x42050a0, 0xc000f30da0}, {0x41c2830, 0xc0014bf3f0}, {0x41ea158, 0xc001472b00}, {0xc001b5bb00, 0x2a}, {0x41ea318, 0xc0012d4730}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper/allocation.go:90 +0x6ba\ngithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper.Keeper.AllocateTokens({{0x42050a0, 0xc000f30da0}, {0x41c2830, 0xc0014bf3f0}, {0x41ea158, 0xc001472b00}, {0xc001b5bb00, 0x2a}, {0x41ea318, 0xc0012d4730}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper/allocation.go:57 +0xd25\ngithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper.EpochsHooksWrapper.AfterEpochEnd({_}, {{0x41ea120, 0x60d2fa0}, {0x42057b0, 0xc002f34c40}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/feedistribution/keeper/hooks.go:38 +0x29e\ngithub.com/ExocoreNetwork/exocore/x/epochs/types.MultiEpochHooks.AfterEpochEnd(...)\n\tgithub.com/ExocoreNetwork/exocore/x/epochs/types/hooks.go:27\ngithub.com/ExocoreNetwork/exocore/x/epochs/keeper.Keeper.BeginBlocker.func1(0xc0059e2780?, {{0xc004b5aff0, 0x6}, {0x4cdea92, 0xede5dda1d, 0x0}, 0xdf8475800, 0x8, {0x4cdea92, 0xede5ddbc1, ...}, ...})\n\tgithub.com/ExocoreNetwork/exocore/x/epochs/keeper/abci.go:82 +0xa3f\ngithub.com/ExocoreNetwork/exocore/x/epochs/keeper.Keeper.IterateEpochInfos({{_, _}, {_, _}, {_, _}}, {{0x41ea120, 0x60d2fa0}, {0x42057b0, 0xc002f34c40}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/x/epochs/keeper/epoch_infos.go:75 +0x1fd\ngithub.com/ExocoreNetwork/exocore/x/epochs/keeper.Keeper.BeginBlocker({{_, _}, {_, _}, {_, _}}, {{0x41ea120, 0x60d2fa0}, {0x42057b0, 0xc002f34c40}, ...})\n\tgithub.com/ExocoreNetwork/exocore/x/epochs/keeper/abci.go:18 +0x206\ngithub.com/ExocoreNetwork/exocore/x/epochs.AppModule.BeginBlock(...)\n\tgithub.com/ExocoreNetwork/exocore/x/epochs/module.go:164\ngithub.com/cosmos/cosmos-sdk/types/module.(*Manager).BeginBlock(_, {{0x41ea120, 0x60d2fa0}, {0x42057b0, 0xc002f34c40}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, ...}, ...}, ...)\n\tgithub.com/cosmos/cosmos-sdk@v0.47.8/types/module/module.go:562 +0x17b\ngithub.com/ExocoreNetwork/exocore/app.(*ExocoreApp).BeginBlocker(_, {{0x41ea120, 0x60d2fa0}, {0x42057b0, 0xc002f34c40}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, ...}, ...}, ...)\n\tgithub.com/ExocoreNetwork/exocore/app/app.go:1166 +0xbb\ngithub.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).BeginBlock(_, {{0xc00203c2e0, 0x20, 0x20}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, {0x20219b84, ...}, ...}, ...})\n\tgithub.com/cosmos/cosmos-sdk@v0.47.8/baseapp/abci.go:196 +0x838\ngithub.com/cometbft/cometbft/abci/client.(*localClient).BeginBlockSync(_, {{0xc00203c2e0, 0x20, 0x20}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, {0x20219b84, ...}, ...}, ...})\n\tgithub.com/cometbft/cometbft@v0.37.4/abci/client/local_client.go:283 +0xfe\ngithub.com/cometbft/cometbft/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc00203c2e0, 0x20, 0x20}, {{0xb, 0x0}, {0xc004530a98, 0x14}, 0x6a, {0x20219b84, ...}, ...}, ...})\n\tgithub.com/cometbft/cometbft@v0.37.4/proxy/app_conn.go:101 +0x195\ngithub.com/cometbft/cometbft/state.execBlockOnProxyApp({0x41ea158?, 0xc0017a83d0}, {0x4201780, 0xc00127e090}, 0xc00550b4a0, {0x4206548, 0xc001785b48}, 0x69?)\n\tgithub.com/cometbft/cometbft@v0.37.4/state/execution.go:363 +0x53d\ngithub.com/cometbft/cometbft/state.(*BlockExecutor).ApplyBlock(_, {{{0xb, 0x0}, {0xc00178e514, 0x6}}, {0xc00175ed80, 0x14}, 0x1, 0x69, {{0xc0053cfdc0, ...}, ...}, ...}, ...)\n\tgithub.com/cometbft/cometbft@v0.37.4/state/execution.go:197 +0x165\ngithub.com/cometbft/cometbft/consensus.(*State).finalizeCommit(0xc000f8e700, 0x6a)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1710 +0xa65\ngithub.com/cometbft/cometbft/consensus.(*State).tryFinalizeCommit(0xc000f8e700, 0x6a)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1619 +0x2f6\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit.func1()\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1554 +0x9c\ngithub.com/cometbft/cometbft/consensus.(*State).enterCommit(0xc000f8e700, 0x6a, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:1592 +0xc5b\ngithub.com/cometbft/cometbft/consensus.(*State).addVote(0xc000f8e700, 0xc002778500, {0x0, 0x0})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2223 +0x18ed\ngithub.com/cometbft/cometbft/consensus.(*State).tryAddVote(0xc000f8e700, 0xc002778500, {0x0?, 0x4fd1f0?})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:2012 +0x26\ngithub.com/cometbft/cometbft/consensus.(*State).handleMsg(0xc000f8e700, {{0x41b7f60, 0xc0000acbd8}, {0x0, 0x0}})\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:865 +0x3ce\ngithub.com/cometbft/cometbft/consensus.(*State).receiveRoutine(0xc000f8e700, 0x0)\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:792 +0x4bc\ncreated by github.com/cometbft/cometbft/consensus.(*State).OnStart in goroutine 1\n\tgithub.com/cometbft/cometbft@v0.37.4/consensus/state.go:383 +0x10c\n"
```

leonz789 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+	// if the rewards do not exist it's fine, we will just add to zero.
+	// allocate share tokens to all stakers of this operator.
+	operatorAccAddress := sdk.AccAddress(valBz)
+	k.AllocateTokensToStakers(ctx, operatorAccAddress, shared, feePool)
```
'AllocateTokenToStakers' seems not properly to be invoked with 'epoch' interval.
It should be either by block or take advantage of 'f1'( f1 is recommandation).
Reason: 'delegation' will change shares of `stakers` under one `oparator` immediately after `tx` is executed successfully which means the portion one `staker` should obtain as reward is possible to be changed on block-level not epoch-level.
eg:
Operator_A has stakers:{a,b} with delegation amount : {10,10}, at the first block in one new epoch, and b delegate 80 more->{10, 90}at the second block, so the final reward should be close to 1:9 for a and b, but if use epoch-distribution for stakers, it would be 1:1.
'F1' will take care of this. we can wrap our multi-tokens as one single 'usd' virtual token, then it can fits into existing 'f1' implementation even for multi-token rewards.
If the 'reward' only take account delegation at epoch-level, it would be pointless for delegation to take effect immediately at tx-level.

mikebraver committed:
fix: fix the negative issue

leonz789 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,151 @@
+)
+
+// Based on the epoch, AllocateTokens performs reward and fee distribution to all validators.
+func (k Keeper) AllocateTokens(ctx sdk.Context, totalPreviousPower int64) error {
```
This totalPreviousPower got from staking/dogfood's lateTotalPower is not actually previous power, it's actually the total power for next epoch, so as the validatorset got from Getallvalidators.
So that means we are distribution the reward from epoch_1 to validatorSet for epoch_2.
example:
1. epoch_1 has validatorSet{a:10,b:90}, totalPower100, received fee of 100exo as reward, so b should take 90(ignore tax here),
2.  but under current implementation ,we will use 'next power info' to do the distribution. Say at end of epoch_1 staking/dogfood update the powers to validatoSet(a:90, c:10), then b will got 0 reward for epoch_1 which is unacceptable.

leonz789 commented:
For the F1 distribution.
existing implementation actually support for 1->n assets diestribution(single staking assets, multiple reward assets).
For our multi-assets staking platform, we can wrap all different kinds of assets into one single assets-USDT, then it will just suit into F1 solution

mikebraver commented:
> For the F1 distribution. existing implementation actually support for 1->n assets diestribution(single staking assets, multiple reward assets). For our multi-assets staking platform, we can wrap all different kinds of assets into one single assets-USDT, then it will just suit into F1 solution

F1 is used for optimization, and for the short-term that the stakers-assets amount can be acceptable to calculate by our current implement, we can further optimize it in next phase

cloud8little commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,164 @@
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(1)
```
curTotalStakersPowers should be 0.

cloud8little commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,164 @@
+				} else {
+					stakersPowerMap[staker] = curStakerPower
+					globalStakerAddressList = append(globalStakerAddressList, staker)
+					curTotalStakersPowers.Add(curStakerPower)
```
```suggestion
					curTotalStakersPowers = curTotalStakersPowers.Add(curStakerPower)
```

mikebraver committed:
fix: fix the power update

cloud8little commented:
Test failed on multiple assets, initialize five validators, two assets for the chain, can notice there is no log for single staker, which means staker don't get reward distributed. 
Notably, there is an outstanding log `10:10AM ERR staker lists not found; skipping module=server` that breaks the distribution process of each staker.
```go
for assetID := range avsAssets {
	stakerList, err := k.StakingKeeper.GetStakersByOperator(ctx, operatorAddress.String(), assetID)
	if err != nil {
		ctx.Logger().Error("staker lists not found; skipping")
	}
	for _, staker := range stakerList.Stakers {
		if curStakerPower, err := k.StakingKeeper.CalculateUSDValueForStaker(ctx, staker, avsAddress, operatorAddress.Bytes()); err != nil {
			ctx.Logger().Error("curStakerPower error", err)
		} else {
			stakersPowerMap[staker] = curStakerPower
			logger.Info("curStakerPower", "is", curStakerPower.String())
			globalStakerAddressList = append(globalStakerAddressList, staker)
			curTotalStakersPowers = curTotalStakersPowers.Add(curStakerPower)
			logger.Info("curTotalStakersPowers", "is", curTotalStakersPowers.String())
		}
	}
}
```

```
 2024/08/28 10:10:33 Triggered, height:6446760, feeder-parames:{feederID:2, startBlock:6456694, interval:30, roundID:1}:
 2024/08/28 10:10:33 Triggered, height:6446760, feeder-parames:{feederID:1, startBlock:6456694, interval:30, roundID:1}:
 10:10AM INF Timed out dur=2987.346754 height=6446761 module=consensus round=0 server=node step=RoundStepNewHeight
 10:10AM INF received proposal module=consensus proposal="Proposal{6446761/0 (7F5F6825DB287D3970F9197642E020A381DB57B11067FC73F1995DD85A837F4D:1:167C6D51AD55, -1) 4522411604A7 @ 2024-08-28T10:10:36.434796761Z}" proposer=9D9F577A10F6C83B928C0268CA3C261CA4412FAE server=node
 10:10AM INF received complete proposal block hash=7F5F6825DB287D3970F9197642E020A381DB57B11067FC73F1995DD85A837F4D height=6446761 module=consensus server=node
 10:10AM INF finalizing commit of block hash=7F5F6825DB287D3970F9197642E020A381DB57B11067FC73F1995DD85A837F4D height=6446761 module=consensus num_txs=0 root=9B0718A6D05C85CB47D29DE4B6452E70DC1FA60E9109032434B73E14246F0609 server=node
 10:10AM INF ending epoch identifier=minute module=x/epochs number=67
 10:10AM INF minted coins from module account amount=20000000000000000000aexo from=exomint module=x/bank
 10:10AM INF AfterEpochEnd minted successfully=20000000000000000000aexo module=x/exomint
 10:10AM INF AfterEpochEnd of distribution module=x/distribution
 10:10AM INF feeCollectorName is=fee_collector module=x/distribution
 10:10AM INF exomint module account is=exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d module=x/distribution
 10:10AM INF distribution module name is=distribution module=x/distribution
 10:10AM INF distribution module account is=exo17xpfvakm2amg962yls6f84z3kell8c5lgqz40d module=x/distribution
 10:10AM INF Allocate tokens to all validators feesCollected amount is =20000000000000000000.000000000000000000aexo module=x/distribution
 10:10AM INF Allocate to stakers of operatorAddress: module=x/distribution
 10:10AM INF operator address is=exo1uw7n7xqxj2hmk5dlw92as6s7se5mmkafuwx982 module=x/distribution
 10:10AM INF assetID is=0x83e6850591425e3c1e263c054f4466838b9bd9e4_0x9ce1 module=x/distribution
 10:10AM INF operator address is=exo1uw7n7xqxj2hmk5dlw92as6s7se5mmkafuwx982 module=x/distribution
 10:10AM INF assetID is=0xb82381a3fbd3fafa77b3a7be693342618240067b_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF allocate tokens to stakers successfully allocated amount is=5134998402969381034.000000000000000000aexo module=x/distribution
 10:10AM INF Allocate tokens to validator successfully allocated amount is=344116216810837414575.681431837329290994aexo module=x/distribution
 10:10AM INF Allocate to stakers of operatorAddress: module=x/distribution
 10:10AM INF operator address is=exo1c5x7mxphvgavjhu0au9jjqnfqcyspevtyy27mz module=x/distribution
 10:10AM INF assetID is=0x83e6850591425e3c1e263c054f4466838b9bd9e4_0x9ce1 module=x/distribution
 10:10AM INF operator address is=exo1c5x7mxphvgavjhu0au9jjqnfqcyspevtyy27mz module=x/distribution
 10:10AM INF assetID is=0xb82381a3fbd3fafa77b3a7be693342618240067b_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF allocate tokens to stakers successfully allocated amount is=5230090965987332529.600000000000000000aexo module=x/distribution
 10:10AM INF Allocate tokens to validator successfully allocated amount is=350488739344371440434.472338826946440838aexo module=x/distribution
 10:10AM INF Allocate to stakers of operatorAddress: module=x/distribution
 10:10AM INF operator address is=exo1u25udsfm5wu3h5s5xqayf6jjycy2etwyfctk5w module=x/distribution
 10:10AM INF assetID is=0x83e6850591425e3c1e263c054f4466838b9bd9e4_0x9ce1 module=x/distribution
 10:10AM INF operator address is=exo1u25udsfm5wu3h5s5xqayf6jjycy2etwyfctk5w module=x/distribution
 10:10AM INF assetID is=0xb82381a3fbd3fafa77b3a7be693342618240067b_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF allocate tokens to stakers successfully allocated amount is=3328239705628302520.600000000000000000aexo module=x/distribution
 10:10AM INF Allocate tokens to validator successfully allocated amount is=223038288673690916758.306893892459444166aexo module=x/distribution
 10:10AM INF Allocate to stakers of operatorAddress: module=x/distribution
 10:10AM INF operator address is=exo1nvgrxlcn854xyrf0jzmrl353gduznvneyassn6 module=x/distribution
 10:10AM INF assetID is=0x83e6850591425e3c1e263c054f4466838b9bd9e4_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF operator address is=exo1nvgrxlcn854xyrf0jzmrl353gduznvneyassn6 module=x/distribution
 10:10AM INF assetID is=0xb82381a3fbd3fafa77b3a7be693342618240067b_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF allocate tokens to stakers successfully allocated amount is=381487396409699813.000000000000000000aexo module=x/distribution
 10:10AM INF Allocate tokens to validator successfully allocated amount is=25564954321623556697.858245512986033479aexo module=x/distribution
 10:10AM INF Allocate to stakers of operatorAddress: module=x/distribution
 10:10AM INF operator address is=exo1lzf5247us2xwhg29mutpx5gyq4aemwqgc08ev7 module=x/distribution
 10:10AM INF assetID is=0x83e6850591425e3c1e263c054f4466838b9bd9e4_0x9ce1 module=x/distribution
 10:10AM INF operator address is=exo1lzf5247us2xwhg29mutpx5gyq4aemwqgc08ev7 module=x/distribution
 10:10AM INF assetID is=0xb82381a3fbd3fafa77b3a7be693342618240067b_0x9ce1 module=x/distribution
 10:10AM ERR staker lists not found; skipping module=server
 10:10AM INF allocate tokens to stakers successfully allocated amount is=5325183529005284044.600000000000000000aexo module=x/distribution
 10:10AM INF Allocate tokens to validator successfully allocated amount is=356861261877905467593.332706844992390641aexo module=x/distribution
 10:10AM INF executed block height=6446761 module=state num_invalid_txs=0 num_valid_txs=0 server=node


```

mikebraver committed:
update

leonz789 commented on x/feedistribution/keeper/hooks.go:
```diff
@@ -0,0 +1,44 @@
+		logger.Info(
+			"AfterEpochEnd of distribution",
+		)
+		err := wrapper.keeper.AllocateTokens(ctx, previousTotalPower.Int64())
```
For 'AllocateTokens', put both totalPower and validatersetPowers as input args, or fetch both of them inside 'AllocateTokens. Seems no special need for fetching totalPower outside AllcateTokens and validatorsetPowers inside 'AllocateTokens'.

mikebraver committed:
update

mikebraver committed:
Merge branch 'develop' into distribution

mikebraver committed:
update

mikebraver committed:
fix: fix action error

leonz789 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,167 @@
+	feesCollected := sdk.NewDecCoinsFromCoins(feesCollectedInt...)
+
+	// transfer collected fees to the distribution module account
+	if err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, types.ModuleName, feesCollectedInt); err != nil {
```
This current implementation of distribution just distribute rewards every epoch instead of using F1 mechanism.
So maybe we can just send the tokens from collector into validators/stakers account as 'withdrawable' value.
And there's no need for 'withdrawReward' then. 
Since this 'record, claim' implementation is used for F1, because the rewards is accumulated when 'withdraw', but in this 'not-f1' version, we just calculate reward every epoch and that value won't change.
Compare:
1. F1: 
  1.1 transfer tokens from collector to distribution(temporary as a vault), setRecords, setPeriodInfomation
  1.2 when delegation/undelegation happened, updatePeriodInformation
  1.3 validator/staker withdraw reward-> calculate finalReward based on above information(recoreds, period). 
  1.4 Send tokens to users.
2. Current implementation without F1
  2.1 the reward is actually fixed every epoch, so we now set in records is actually the final rewards need no more calculation and period information.
  2.2 then if we don't use F1, it could just be: transfer tokens directly from collector->validator/stakers.
*(Though F1 is recommanded)*

mikebraver replied:
Yes, the struct keeps compatible for now for easier further optimized

cloud8little committed:
Merge branch 'develop' into distribution

mikebraver committed:
Merge branch 'develop' into distribution

mikebraver committed:
update

mikebraver committed:
Merge branch 'develop' into distribution

cloud8little commented on x/reward/keeper/keeper.go:
```diff
@@ -5,33 +5,81 @@ import (
+	store.Set(poolKey.Bytes(), k.cdc.MustMarshal(&pool))
+}
+
+// TODO: to be enabled later by avs
+// func (k Keeper) getPools(ctx sdk.Context) ([]types.Pool, error) {
+//	var pools []types.Pool
+//
+//	poolNamePrefix := utils.LowerCaseKey(poolNamePrefix)
+//	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixRewardInfo)
+//	iter := sdk.KVStorePrefixIterator(store, append(poolNamePrefix.AsKey(), []byte(DefaultDelimiter)...))
+//	defer utils.CloseLogError(iter, k.Logger(ctx))
+//
+//	for ; iter.Valid(); iter.Next() {
+//		var pool types.Pool
+//		k.cdc.MustUnmarshal(iter.Value(), &pool)
+//		pools = append(pools, pool)
+//	}
+//
+//	return pools, nil
+//}
```
yes.

cloud8little commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,171 @@
+			stakerPower := stakersPowerMap[staker]
+			powerFraction := stakerPower.QuoTruncate(curTotalStakersPowers)
+			rewardToSingleStaker := rewardToAllStakers.MulDecTruncate(powerFraction)
+			if rw := rewardToAllStakers.Sub(rewardToSingleStaker); rw.IsValid() {
```
the logic here wrong, rewardToAllStakers value should **NOT** be changed, another variable like remaining should be defined and initialize with rewardToAllStakers, then for each staker, after calculating the reward for the staker, update the remaining reward value to check if all rewards are distributed.

mikebraver committed:
fix: fix staker allocations

MaxMustermann2 committed:
fix(dogfood): correct audit typos (#190)

No other recommendations were received for the x/dogfood module.

MaxMustermann2 committed:
feat(assets): remove total supply (#181)

* feat(assets): validate total supply before edit

Now that the TVL limits are imposed on the client chain instead of
Exocore, this PR renames the variable `tvlLimit` to `totalSupply`. In
addition, editing the `totalSupply` to a value lower than the currently
deposited tokens is not permitted.

* fix(doc): update IAssets comment

* feat: introduce `getTotalSupply` precompile

This can be used to fetch the total supply, according to Exocore, for an
asset.

* fix: remove staking asset total supply

The `totalSupply` variable was created to guard against black swan
events wherein a malicious transaction could try to deposit more than
the quantity of tokens that exist. The benefits of this feature are
negligible, and the costs to maintain the constant updates between the
"real" total supply on the client chain and the "recorded" total supply
on Exocore are too high. Hence, it has been removed in its entirety.

Instead, TVL limits will be imposed by the client chain for an
operational risk management perspective.

* fix the lint error caused by the lower curl version

---------

Co-authored-by: TimmyExogenous <chuang@exogenous.io>

mikebraver committed:
update

cloud8little commented on x/feedistribution/keeper/hooks_test.go:
```diff
@@ -41,6 +42,10 @@ func (suite *KeeperTestSuite) TestEpochHooks() {
 			continue
 		}
 		valBz := validatorDetail.GetOperator()
+		// accAddr := sdk.AccAddress(valBz)
```
remove the debug code plz.

mikebraver committed:
lint: fix github action error

mikebraver committed:
Merge branch 'develop' into distribution

mikebraver committed:
lint

mikebraver committed:
rebase

mikebraver committed:
resolve: some updates resolveing comments

mikebraver committed:
merge: merge develop

mikebraver committed:
merge: merge develop

mikebraver committed:
update

MaxMustermann2 commented on local_node.sh:
```diff
@@ -126,6 +126,13 @@ if [[ $overwrite == "y" || $overwrite == "Y" ]]; then
 	jq '.app_state["oracle"]["params"]["token_feeders"][1]["start_base_block"]="20"' "$GENESIS" >"$TMP_GENESIS" && mv "$TMP_GENESIS" "$GENESIS"
 
 	# x/operator
+<<<<<<< HEAD
```
Conflict

MaxMustermann2 commented on x/feedistribution/client/cli/query.go:
```diff
@@ -0,0 +1,48 @@
+	return cmd
+}
+
+func QueryDistributionInfo() *cobra.Command {
```
```suggestion
func CmdQueryParams() *cobra.Command {
```

MaxMustermann2 commented on x/feedistribution/client/cli/query.go:
```diff
@@ -0,0 +1,48 @@
+		RunE:                       client.ValidateCmd,
+	}
+
+	cmd.AddCommand(QueryDistributionInfo())
```
```suggestion
	cmd.AddCommand(CmdQueryParams())
```

MaxMustermann2 commented on x/feedistribution/client/cli/query.go:
```diff
@@ -0,0 +1,48 @@
+
+func QueryDistributionInfo() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "DistributionInfo query",
+		Short: "DistributionInfo query",
```
```suggestion
		Use:   "params",
		Short: "show module params",
```

MaxMustermann2 commented on x/feedistribution/client/cli/tx.go:
```diff
@@ -0,0 +1,74 @@
+					CommunityTax:    communityTax,
+				},
+			}
+			if err := msg.ValidateBasic(); err != nil {
```
Not necessary since `GenerateOrBroadcastTxCLI` calls this

MaxMustermann2 commented on x/feedistribution/types/errors.go:
```diff
@@ -0,0 +1,13 @@
+// x/feedistribution module sentinel errors
+var (
+	ErrInvalidSigner = sdkerrors.Register(ModuleName, 1100, "expected gov account as only signer for proposal message")
+	ErrSample        = sdkerrors.Register(ModuleName, 1101, "sample error")
```
remove this please

MaxMustermann2 commented on x/feedistribution/types/msg_update_params.go:
```diff
@@ -0,0 +1,32 @@
+		return errorsmod.Wrap(err, "invalid authority address")
+	}
+
+	if err := m.Params.Validate(); err != nil {
```
This is only stateless validation. When processing the message, stateful validation must be performed: check whether the provided epoch is registered in x/epochs.

MaxMustermann2 commented on x/feedistribution/types/params.go:
```diff
@@ -0,0 +1,44 @@
+	return paramtypes.ParamSetPairs{}
+}
+
+// Validate validates the set of params
+func (p Params) Validate() error {
+	return nil
+}
```
Agree with this one

MaxMustermann2 commented on x/feedistribution/keeper/genesis.go:
```diff
@@ -0,0 +1,28 @@
+	if !found {
+		// the panic is suitable here because it is being done at genesis, when the node
+		// is not running. it means that the genesis file is malformed.
+		log.Error()
```
`panic` here

mikebraver committed:
resolve: resolve further comments

MaxMustermann2 commented on proto/exocore/feedistribution/v1/params.proto:
```diff
@@ -0,0 +1,22 @@
+
+// Params defines the parameters for the module.
+message Params {
+  option (amino.name) = "exocore/x/feedistribution/Params";
```
Not a blocker, but what does this do? I haven't seen it before.

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,171 @@
+			logger.Error("Operator address not found; skipping", "consAddress", sdk.GetConsAddress(pk), "i", i)
+			continue
+		}
+		if totalPreviousPower == 0 {
+			return nil
+		}
```
```suggestion
```
This can be removed since we already checked this outside the loop. Or did you mean to check `val.Power == 0` ?

MaxMustermann2 commented on x/feedistribution/keeper/allocation.go:
```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		ctx.Logger().Error("avs address lists not found; skipping")
+	}
+	stakersPowerMap, curTotalStakersPowers := make(map[string]math.LegacyDec), math.LegacyNewDec(0)
```
Recheck this @coderabbitai

MaxMustermann2 commented on x/feedistribution/types/errors.go:
```diff
@@ -0,0 +1,15 @@
+// x/feedistribution module sentinel errors
+var (
+	ErrEpochNotFound = sdkerrors.Register(
+		ModuleName, 1102,
```
Any reason to start with 1102 and not 2?

MaxMustermann2 commented on x/feedistribution/types/params.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+// Validate validates the set of params
+func (p Params) Validate() error {
+	return nil
+}
```
```suggestion
func (p Params) Validate() error {
	if err := epochstypes.ValidateEpochIdentifierString(p.EpochIdentifier); err != nil {
	    return fmt.Errorf("invalid epoch ID %s", err)
	}
	if p.CommunityTax.IsNil() || p.CommunityTax.IsNegative() || p.CommunityTax.GT(math.LegacyOneDec()) {
	    return fmt.Errorf("invalid community tax %d", p.CommunityTax)
	}
	return nil
}
```

MaxMustermann2 commented:
Minor changes suggested; looks good.

mikebraver committed:
resolve: some updates

mikebraver committed:
resolve: minor update
